src/useForm.tsx

```diff
@@ -74,19 +74,28 @@ function createFields<T>(
           value,
           error,
           touched,
+          // For onChange / onBlur, we use a "functional" setState call
+          // (https://reactjs.org/docs/hooks-reference.html#functional-updates)
+          // This avoids issues related to stale state when multiple onChange functions are
+          // invoked within the same render cycle -- e.g.
+          // const onClick = () => { date.day.onChange(...); date.day.onChange(...) }:
           onChange: (updatedValue: any) => {
-            const updatedState = produce(state, updatedState => {
-              set(updatedState, [...path, 'value'], updatedValue)
-              set(updatedState, [...path, 'touched'], true)
-            })
-            setState(updatedState)
+            setState(currentState =>
+              produce(currentState, updatedState => {
```

Comment:
This `produce` function is from [Immer](https://github.com/immerjs/immer). Basically it gives you an efficient way to perform immutable updates on JS objects.

And under the covers it generates a proxy object (`updatedState` here), observes what you do with it, and then captures the diff.